tsl/src/remote/connection.c

```diff
@@ -1390,6 +1408,60 @@ add_userinfo_to_server_options(ForeignServer *server, Oid user_id)
 	return options;
 }
 
+/*
+ * Escaping libpq connect parameter strings.
+ *
+ * Replaces "'" with "\'" and "\" with "\\".
+ */
+static char *
+escape_param_str(const char *str)
+{
+	const char *cp;
+	StringInfoData buf;
+
+	initStringInfo(&buf);
+
+	for (cp = str; *cp; cp++)
+	{
+		if (*cp == '\\' || *cp == '\'')
+			appendStringInfoChar(&buf, '\\');
+		appendStringInfoChar(&buf, *cp);
+	}
+
+	return buf.data;
+}
+
+char *
+remote_connection_get_connstr(const char *node_name)
+{
+	ForeignServer *server;
+	List *connection_options;
+	const char **keywords;
+	const char **values;
+	StringInfoData buf;
+	int i;
+
+	server = data_node_get_foreign_server(node_name, ACL_NO_CHECK, false, false);
+	connection_options = add_userinfo_to_server_options(server, GetUserId());
+	setup_full_connection_options(connection_options, &keywords, &values);
+
+	/* Cycle through the options and create the connection string */
+	initStringInfo(&buf);
+	i = 0;
+	while (keywords[i] != NULL)
+	{
+		appendStringInfo(&buf, "%s=%s ", keywords[i], escape_param_str(values[i]));
```

Comment:
Might be a good idea to call pfree() here on the string returned by escape_param_str()